File: app/src/main/java/maderski/bluetoothautoplaymusic/UI/HeadphonesFragment.java
package maderski.bluetoothautoplaymusic.UI;

import android.content.Context;
import android.content.res.ColorStateList;
import android.graphics.Typeface;
import android.media.AudioManager;
import android.os.Bundle;
import android.support.v4.app.DialogFragment;
import android.support.v4.content.ContextCompat;
import android.support.v4.widget.CompoundButtonCompat;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.LinearLayout;
import android.widget.SeekBar;
import android.widget.TextView;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

import maderski.bluetoothautoplaymusic.Helpers.BluetoothDeviceHelper;
import maderski.bluetoothautoplaymusic.BuildConfig;
import maderski.bluetoothautoplaymusic.R;
import maderski.bluetoothautoplaymusic.SharedPrefs.BAPMPreferences;

public class HeadphonesFragment extends DialogFragment{

    private static final String TAG = "HeadphonesFragment";

    private HashSet<String> removedDevices = new HashSet<>();

    private OnFragmentInteractionListener mListener;

    public HeadphonesFragment() {
        // Required empty public constructor
    }

    public static HeadphonesFragment newInstance() {
        HeadphonesFragment fragment = new HeadphonesFragment();
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View rootView = inflater.inflate(R.layout.fragment_headphones, container, false);
        checkboxCreator(rootView);

        AudioManager audioManager = (AudioManager)getActivity().getSystemService(Context.AUDIO_SERVICE);
        SeekBar volumeSeekBar = (SeekBar)rootView.findViewById(R.id.volume_seekBar);
        volumeSeekBar.setMax(audioManager.getStreamMaxVolume(AudioManager.STREAM_MUSIC));
        volumeSeekBar.setProgress(BAPMPreferences.getHeadphonePreferredVolume(getActivity()));
        volumeSeekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                BAPMPreferences.setHeadphonePreferredVolume(getActivity(), progress);
                Log.d(TAG, "Progress: " + Integer.toString(progress));
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {

            }
        });

        Button doneButton = (Button)rootView.findViewById(R.id.autoplay_done);
        doneButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mListener.headphonesDoneClicked(removedDevices);
                dismiss();
            }
        });
        if(removedDevices != null)
            removedDevices.clear();
        return rootView;
    }

    private Set<String> getNonHeadphoneDevices(){
        Set<String> btDevices = BAPMPreferences.getBTDevices(getContext());
        Set<String> headphoneDevices = BAPMPreferences.getHeadphoneDevices(getContext());

        for(String headphoneDevice : headphoneDevices){
            if(btDevices.contains(headphoneDevice)){
                btDevices.remove(headphoneDevice);
            }
        }

        return btDevices;
    }

    //Create Checkboxes
    private void checkboxCreator(View view) {

        CheckBox checkBox;
        TextView textView;

        LinearLayout autoplayCkBoxLL = (LinearLayout) view.findViewById(R.id.autoplay_only_ll);
        autoplayCkBoxLL.removeAllViews();
        List<String> listOfBTDevices = BluetoothDeviceHelper.listOfBluetoothDevices();
        if (listOfBTDevices.contains("No Bluetooth Device found") ||
                listOfBTDevices.isEmpty()){
            textView = new TextView(getActivity());
            textView.setText(R.string.no_BT_found);
            autoplayCkBoxLL.addView(textView);
        }else{
            for (String BTDevice : listOfBTDevices) {
                int textColor = R.color.colorPrimary;
                checkBox = new CheckBox(getActivity());
                checkBox.setText(BTDevice);
                if(getNonHeadphoneDevices().contains(BTDevice)) {
                    textColor = R.color.lightGray;
                    int states[][] = {{android.R.attr.state_checked}};
                    int colors[] = {textColor, textColor};
                    CompoundButtonCompat.setButtonTintList(checkBox, new ColorStateList(states, colors));
                    checkBox.setClickable(false);
                    checkBox.setChecked(true);
                } else if(BAPMPreferences.getBTDevices(getActivity()) != null) {
                    checkBox.setChecked(BAPMPreferences.getHeadphoneDevices(getContext()).contains(BTDevice));
                }
                checkBox.setTextColor(ContextCompat.getColor(getContext(), textColor));
                checkBox.setTypeface(Typeface.createFromAsset(getActivity().getAssets(), "fonts/TitilliumText400wt.otf"));

                if(!getNonHeadphoneDevices().contains(BTDevice)) {
                    checkboxListener(view.getContext(), checkBox, BTDevice);
                }
                autoplayCkBoxLL.addView(checkBox);
            }
        }

    }

    //Get Selected Checkboxes
    private void checkboxListener(final Context context, CheckBox checkBox, String BTDevice) {
        final CheckBox cb = checkBox;
        final String BTD = BTDevice;

        cb.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                HashSet<String> savedHeadphoneDevices = new HashSet<>(BAPMPreferences.getHeadphoneDevices(context));
                if (cb.isChecked()) {
                    savedHeadphoneDevices.add(BTD);
                    if(removedDevices.contains(BTD)){
                        removedDevices.remove(BTD);
                    }
                    if (BuildConfig.DEBUG) {
                        Log.i(TAG, "TRUE" + " " + BTD);
                        Log.i(TAG, "SAVED");
                    }
                } else {
                    savedHeadphoneDevices.remove(BTD);
                    removedDevices.add(BTD);
                    if (BuildConfig.DEBUG)
                        Log.i(TAG, "FALSE" + " " + BTD);
                    if (BuildConfig.DEBUG)
                        Log.i(TAG, "SAVED");
                }
                mListener.setHeadphoneDevices(savedHeadphoneDevices);
                mListener.headDeviceSelection(BTD, cb.isChecked());
            }
        });
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnFragmentInteractionListener) {
            mListener = (OnFragmentInteractionListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnFragmentInteractionListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     * <p>
     * See the Android Training lesson <a href=
     * "http://developer.android.com/training/basics/fragments/communicating.html"
     * >Communicating with Other Fragments</a> for more information.
     */
    public interface OnFragmentInteractionListener {
        void setHeadphoneDevices(HashSet<String> headphoneDevices);
        void headphonesDoneClicked(HashSet<String> removedDevices);
        void headDeviceSelection(String deviceName, boolean addDevice);
    }
}

File: app/src/main/java/maderski/bluetoothautoplaymusic/ui/fragments/HeadphonesFragment.kt
package maderski.bluetoothautoplaymusic.ui.fragments

import android.content.Context
import android.content.res.ColorStateList
import android.graphics.Typeface
import android.media.AudioManager
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.core.content.ContextCompat
import androidx.core.widget.CompoundButtonCompat
import maderski.bluetoothautoplaymusic.BuildConfig
import maderski.bluetoothautoplaymusic.R
import maderski.bluetoothautoplaymusic.bluetooth.models.BAPMDevice
import maderski.bluetoothautoplaymusic.sharedprefs.BAPMPreferences
import maderski.bluetoothautoplaymusic.helpers.BluetoothDeviceHelper
import maderski.bluetoothautoplaymusic.wrappers.SystemServicesWrapper
import org.koin.android.ext.android.inject
import java.util.*

class HeadphonesFragment : androidx.fragment.app.DialogFragment() {
    private val preferences: BAPMPreferences by inject()
    private val systemServicesWrapper: SystemServicesWrapper by inject()
    private val bluetoothDeviceHelper: BluetoothDeviceHelper by inject()

    private val removedDevices = mutableSetOf<BAPMDevice>()
    private val nonHeadphoneDevices:  Set<BAPMDevice>
        get() {
            val btDevices = preferences.getBAPMDevices().toMutableSet()
            val headphoneDevices = preferences.getHeadphoneDevices()

            btDevices.removeAll(headphoneDevices)

            return btDevices
        }

    private var fragmentInteractionListener: OnFragmentInteractionListener? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        val rootView = inflater.inflate(R.layout.fragment_headphones, container, false)
        checkboxCreator(rootView)

        val audioManager = systemServicesWrapper.audioManager
        val volumeSeekBar = rootView.findViewById<View>(R.id.volume_seekBar) as SeekBar
        volumeSeekBar.max = audioManager.getStreamMaxVolume(AudioManager.STREAM_MUSIC)
        volumeSeekBar.progress = preferences.getHeadphonePreferredVolume()
        volumeSeekBar.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                preferences.setHeadphonePreferredVolume(progress)
                Log.d(TAG, "Progress: $progress")
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {

            }

            override fun onStopTrackingTouch(seekBar: SeekBar) {

            }
        })

        val doneButton = rootView.findViewById<View>(R.id.autoplay_done) as Button
        doneButton.setOnClickListener {
            fragmentInteractionListener?.headphonesDoneClicked(removedDevices)
            dismiss()
        }

        val a2dpSwitch = rootView.findViewById<View>(R.id.sw_headphones_a2dp) as Switch
        val useA2dp = preferences.getUseA2dpHeadphones()
        a2dpSwitch.isChecked = useA2dp
        a2dpSwitch.setOnCheckedChangeListener { _, isChecked -> fragmentInteractionListener?.onUseHeadphonesA2DP(isChecked) }

        removedDevices.clear()
        return rootView
    }

    //Create Checkboxes
    private fun checkboxCreator(view: View) {

        var checkBox: CheckBox
        val textView: TextView

        val autoplayCkBoxLL = view.findViewById<View>(R.id.autoplay_only_ll) as LinearLayout
        autoplayCkBoxLL.removeAllViews()
        val listOfBTDevices = bluetoothDeviceHelper.listOfBluetoothDevices()
        if (listOfBTDevices.isEmpty()) {
            textView = TextView(activity)
            textView.setText(R.string.no_BT_found)
            autoplayCkBoxLL.addView(textView)
        } else {
            for (bapmDevice in listOfBTDevices) {
                var textColor = R.color.colorPrimary
                checkBox = CheckBox(activity)
                checkBox.text = bapmDevice.name
                if (nonHeadphoneDevices.contains(bapmDevice)) {
                    textColor = R.color.lightGray
                    val states = arrayOf(intArrayOf(android.R.attr.state_checked))
                    val colors = intArrayOf(textColor, textColor)
                    CompoundButtonCompat.setButtonTintList(checkBox, ColorStateList(states, colors))
                    checkBox.isClickable = false
                    checkBox.isChecked = true
                } else if (preferences.getBAPMDevices().isNotEmpty()) {
                    checkBox.isChecked = preferences.getHeadphoneDevices().contains(bapmDevice)
                }
                checkBox.setTextColor(ContextCompat.getColor(requireActivity(), textColor))
                checkBox.typeface = Typeface.createFromAsset(requireActivity().assets, "fonts/TitilliumText400wt.otf")

                if (!nonHeadphoneDevices.contains(bapmDevice)) {
                    checkboxListener(checkBox, bapmDevice)
                }
                autoplayCkBoxLL.addView(checkBox)
            }
        }

    }

    //Get Selected Checkboxes
    private fun checkboxListener(checkBox: CheckBox, bapmDevice: BAPMDevice) {

        checkBox.setOnClickListener {
            val savedHeadphoneDevices = preferences.getHeadphoneDevices().toMutableSet()
            if (checkBox.isChecked) {
                savedHeadphoneDevices.add(bapmDevice)
                if (removedDevices.contains(bapmDevice)) {
                    removedDevices.remove(bapmDevice)
                }
                if (BuildConfig.DEBUG) {
                    Log.i(TAG, "TRUE $bapmDevice")
                    Log.i(TAG, "SAVED")
                }
            } else {
                savedHeadphoneDevices.remove(bapmDevice)
                removedDevices.add(bapmDevice)
                if (BuildConfig.DEBUG)
                    Log.i(TAG, "FALSE $bapmDevice")
                if (BuildConfig.DEBUG)
                    Log.i(TAG, "SAVED")
            }
            fragmentInteractionListener?.setHeadphoneDevices(savedHeadphoneDevices)
            fragmentInteractionListener?.headDeviceSelection(bapmDevice.name, checkBox.isChecked)
        }
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is OnFragmentInteractionListener) {
            fragmentInteractionListener = context
        } else {
            throw RuntimeException("$context must implement OnFragmentInteractionListener")
        }
    }

    override fun onDetach() {
        super.onDetach()
        fragmentInteractionListener = null
    }

    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     *
     *
     * See the Android Training lesson [Communicating with Other Fragments](http://developer.android.com/training/basics/fragments/communicating.html) for more information.
     */
    interface OnFragmentInteractionListener {
        fun setHeadphoneDevices(headphoneDevices:  Set<BAPMDevice>)
        fun headphonesDoneClicked(removedDevices:  Set<BAPMDevice>)
        fun headDeviceSelection(deviceName: String, addDevice: Boolean)
        fun onUseHeadphonesA2DP(isUsingA2DP: Boolean)
    }

    companion object {

        private const val TAG = "HeadphonesFragment"

        fun newInstance(): HeadphonesFragment {
            return HeadphonesFragment()
        }
    }
}
